File: IOS/Sources/TransferCenter.swift
import Foundation

enum TransferDirection: String, Codable {
    case send
    case receive
}

struct TransferSnapshot: Identifiable {
    let id: String
    let direction: TransferDirection
    let fileName: String
    let peer: String
    let bytesTransferred: Int64
    let totalBytes: Int64
    let progress: Double?
    let bytesPerSecond: Double
    let etaSeconds: Int64?
    let startedAt: Int64
}

struct TransferHistoryEntry: Identifiable, Codable {
    let id: String
    let direction: TransferDirection
    let fileName: String
    let peer: String
    let bytesTransferred: Int64
    let totalBytes: Int64
    let startedAt: Int64
    let finishedAt: Int64
    let success: Bool
    let detail: String?
}

private final class MutableTransfer {
    let id: String
    let direction: TransferDirection
    let fileName: String
    let peer: String
    let totalBytes: Int64
    let startedAt: Int64
    var bytesTransferred: Int64 = 0
    var lastSampleAt: Int64
    var lastSampleBytes: Int64 = 0
    var smoothedBytesPerSecond: Double = 0

    init(id: String, direction: TransferDirection, fileName: String, peer: String, totalBytes: Int64, startedAt: Int64) {
        self.id = id
        self.direction = direction
        self.fileName = fileName
        self.peer = peer
        self.totalBytes = totalBytes
        self.startedAt = startedAt
        self.lastSampleAt = startedAt
    }
}

final class TransferCenter {
    static let shared = TransferCenter()
    private let lock = NSLock()
    private var activeById: [String: MutableTransfer] = [:]
    private var historyEntries: [TransferHistoryEntry] = []
    private let historyKey = "ftportal.transfer.history.v1"
    private let maximumHistory = 100

    private init() {
        if let data = UserDefaults.standard.data(forKey: historyKey),
           let decoded = try? JSONDecoder().decode([TransferHistoryEntry].self, from: data) {
            historyEntries = Array(decoded.prefix(maximumHistory))
        }
    }

    func begin(direction: TransferDirection, fileName: String, peer: String, totalBytes: Int64) -> String {
        let id = UUID().uuidString
        let now = Self.nowMilliseconds()
        let transfer = MutableTransfer(
            id: id,
            direction: direction,
            fileName: fileName.isEmpty ? "Unnamed file" : fileName,
            peer: peer.isEmpty ? "Local network peer" : peer,
            totalBytes: totalBytes,
            startedAt: now
        )
        lock.lock()
        activeById[id] = transfer
        lock.unlock()
        return id
    }

    func update(id: String, bytes: Int64) {
        let now = Self.nowMilliseconds()
        lock.lock()
        defer { lock.unlock() }
        guard let transfer = activeById[id] else { return }
        let absolute = max(0, bytes)
        let deltaBytes = absolute - transfer.lastSampleBytes
        let deltaMs = now - transfer.lastSampleAt
        transfer.bytesTransferred = absolute
        if deltaBytes >= 0 && deltaMs >= 180 {
            let instant = Double(deltaBytes) * 1000 / Double(deltaMs)
            transfer.smoothedBytesPerSecond = transfer.smoothedBytesPerSecond <= 0
                ? instant
                : transfer.smoothedBytesPerSecond * 0.72 + instant * 0.28
            transfer.lastSampleAt = now
            transfer.lastSampleBytes = absolute
        }
    }

    func finish(id: String, success: Bool, detail: String? = nil) {
        lock.lock()
        guard let transfer = activeById.removeValue(forKey: id) else {
            lock.unlock()
            return
        }
        let entry = TransferHistoryEntry(
            id: transfer.id,
            direction: transfer.direction,
            fileName: transfer.fileName,
            peer: transfer.peer,
            bytesTransferred: transfer.bytesTransferred,
            totalBytes: transfer.totalBytes,
            startedAt: transfer.startedAt,
            finishedAt: Self.nowMilliseconds(),
            success: success,
            detail: detail.map { String($0.prefix(240)) }
        )
        historyEntries.insert(entry, at: 0)
        if historyEntries.count > maximumHistory {
            historyEntries.removeLast(historyEntries.count - maximumHistory)
        }
        let historyCopy = historyEntries
        lock.unlock()
        persist(historyCopy)
    }

    func active() -> [TransferSnapshot] {
        lock.lock()
        let result = activeById.values.map { transfer -> TransferSnapshot in
            let total = transfer.totalBytes
            let bytes = transfer.bytesTransferred
            let speed = transfer.smoothedBytesPerSecond
            let progress = total > 0 ? min(1, Double(bytes) / Double(total)) : nil
            let eta: Int64? = total > 0 && speed > 1 && bytes < total
                ? max(0, Int64(Double(total - bytes) / speed))
                : nil
            return TransferSnapshot(
                id: transfer.id,
                direction: transfer.direction,
                fileName: transfer.fileName,
                peer: transfer.peer,
                bytesTransferred: bytes,
                totalBytes: total,
                progress: progress,
                bytesPerSecond: speed,
                etaSeconds: eta,
                startedAt: transfer.startedAt
            )
        }.sorted { $0.startedAt < $1.startedAt }
        lock.unlock()
        return result
    }

    func history() -> [TransferHistoryEntry] {
        lock.lock()
        let result = historyEntries
        lock.unlock()
        return result
    }

    func clearHistory() {
        lock.lock()
        historyEntries.removeAll()
        lock.unlock()
        UserDefaults.standard.removeObject(forKey: historyKey)
    }

    private func persist(_ entries: [TransferHistoryEntry]) {
        guard let data = try? JSONEncoder().encode(entries) else { return }
        UserDefaults.standard.set(data, forKey: historyKey)
    }

    private static func nowMilliseconds() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}

File: ANDROID/app/src/main/java/com/mrhakan/ftportal/TransferCenter.kt
package com.mrhakan.ftportal

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.max
import kotlin.math.min

enum class TransferDirection { SEND, RECEIVE }

data class TransferSnapshot(
    val id: String,
    val direction: TransferDirection,
    val fileName: String,
    val peer: String,
    val bytesTransferred: Long,
    val totalBytes: Long,
    val progressPercent: Int?,
    val bytesPerSecond: Double,
    val etaSeconds: Long?,
    val startedAt: Long
)

data class TransferHistoryEntry(
    val id: String,
    val direction: TransferDirection,
    val fileName: String,
    val peer: String,
    val bytesTransferred: Long,
    val totalBytes: Long,
    val startedAt: Long,
    val finishedAt: Long,
    val success: Boolean,
    val detail: String?
)

private data class MutableTransfer(
    val id: String,
    val direction: TransferDirection,
    val fileName: String,
    val peer: String,
    val totalBytes: Long,
    val startedAt: Long,
    var bytesTransferred: Long = 0L,
    var lastSampleAt: Long = startedAt,
    var lastSampleBytes: Long = 0L,
    var smoothedBytesPerSecond: Double = 0.0
)

object TransferCenter {
    private const val PREFS = "ftportal-transfer-history"
    private const val KEY_HISTORY = "history"
    private const val MAX_HISTORY = 100

    private val active = ConcurrentHashMap<String, MutableTransfer>()
    private val historyLock = Any()
    @Volatile private var cachedHistory: MutableList<TransferHistoryEntry>? = null

    fun initialize(context: Context) {
        if (cachedHistory != null) return
        synchronized(historyLock) {
            if (cachedHistory == null) cachedHistory = loadHistory(context.applicationContext)
        }
    }

    fun begin(
        context: Context,
        direction: TransferDirection,
        fileName: String,
        peer: String,
        totalBytes: Long
    ): String {
        initialize(context)
        val id = UUID.randomUUID().toString()
        val now = System.currentTimeMillis()
        active[id] = MutableTransfer(
            id = id,
            direction = direction,
            fileName = fileName.ifBlank { "Unnamed file" },
            peer = peer.ifBlank { "Local network peer" },
            totalBytes = totalBytes,
            startedAt = now
        )
        return id
    }

    fun update(id: String, absoluteBytes: Long) {
        val transfer = active[id] ?: return
        val now = System.currentTimeMillis()
        synchronized(transfer) {
            val bytes = max(0L, absoluteBytes)
            val deltaBytes = bytes - transfer.lastSampleBytes
            val deltaMs = now - transfer.lastSampleAt
            transfer.bytesTransferred = bytes
            if (deltaBytes >= 0L && deltaMs >= 180L) {
                val instant = deltaBytes.toDouble() * 1000.0 / deltaMs.toDouble()
                transfer.smoothedBytesPerSecond = if (transfer.smoothedBytesPerSecond <= 0.0) {
                    instant
                } else {
                    transfer.smoothedBytesPerSecond * 0.72 + instant * 0.28
                }
                transfer.lastSampleAt = now
                transfer.lastSampleBytes = bytes
            }
        }
    }

    fun finish(context: Context, id: String, success: Boolean, detail: String? = null) {
        initialize(context)
        val transfer = active.remove(id) ?: return
        val snapshot = synchronized(transfer) {
            TransferHistoryEntry(
                id = transfer.id,
                direction = transfer.direction,
                fileName = transfer.fileName,
                peer = transfer.peer,
                bytesTransferred = transfer.bytesTransferred,
                totalBytes = transfer.totalBytes,
                startedAt = transfer.startedAt,
                finishedAt = System.currentTimeMillis(),
                success = success,
                detail = detail?.take(240)
            )
        }
        synchronized(historyLock) {
            val history = cachedHistory ?: mutableListOf<TransferHistoryEntry>().also { cachedHistory = it }
            history.add(0, snapshot)
            while (history.size > MAX_HISTORY) history.removeAt(history.lastIndex)
            persistHistory(context.applicationContext, history)
        }
    }

    fun active(): List<TransferSnapshot> = active.values.map { transfer ->
        synchronized(transfer) {
            val total = transfer.totalBytes
            val bytes = transfer.bytesTransferred
            val speed = transfer.smoothedBytesPerSecond
            val percent = if (total > 0L) {
                ((min(bytes, total).toDouble() / total.toDouble()) * 100.0).toInt().coerceIn(0, 100)
            } else null
            val eta = if (total > 0L && speed > 1.0 && bytes < total) {
                ((total - bytes).toDouble() / speed).toLong().coerceAtLeast(0L)
            } else null
            TransferSnapshot(
                id = transfer.id,
                direction = transfer.direction,
                fileName = transfer.fileName,
                peer = transfer.peer,
                bytesTransferred = bytes,
                totalBytes = total,
                progressPercent = percent,
                bytesPerSecond = speed,
                etaSeconds = eta,
                startedAt = transfer.startedAt
            )
        }
    }.sortedBy { it.startedAt }

    fun history(context: Context): List<TransferHistoryEntry> {
        initialize(context)
        synchronized(historyLock) { return cachedHistory.orEmpty().toList() }
    }

    fun clearHistory(context: Context) {
        initialize(context)
        synchronized(historyLock) {
            cachedHistory?.clear()
            context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
                .edit().remove(KEY_HISTORY).apply()
        }
    }

    private fun loadHistory(context: Context): MutableList<TransferHistoryEntry> {
        val raw = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).getString(KEY_HISTORY, null)
            ?: return mutableListOf()
        val array = runCatching { JSONArray(raw) }.getOrNull() ?: return mutableListOf()
        val result = mutableListOf<TransferHistoryEntry>()
        for (index in 0 until min(array.length(), MAX_HISTORY)) {
            val item = array.optJSONObject(index) ?: continue
            val direction = runCatching { TransferDirection.valueOf(item.optString("direction")) }.getOrNull() ?: continue
            result += TransferHistoryEntry(
                id = item.optString("id").ifBlank { UUID.randomUUID().toString() },
                direction = direction,
                fileName = item.optString("fileName").ifBlank { "Unnamed file" },
                peer = item.optString("peer").ifBlank { "Local network peer" },
                bytesTransferred = item.optLong("bytesTransferred", 0L),
                totalBytes = item.optLong("totalBytes", -1L),
                startedAt = item.optLong("startedAt", 0L),
                finishedAt = item.optLong("finishedAt", 0L),
                success = item.optBoolean("success", false),
                detail = item.optString("detail").takeIf { it.isNotBlank() }
            )
        }
        return result
    }

    private fun persistHistory(context: Context, history: List<TransferHistoryEntry>) {
        val array = JSONArray()
        history.take(MAX_HISTORY).forEach { entry ->
            array.put(
                JSONObject()
                    .put("id", entry.id)
                    .put("direction", entry.direction.name)
                    .put("fileName", entry.fileName)
                    .put("peer", entry.peer)
                    .put("bytesTransferred", entry.bytesTransferred)
                    .put("totalBytes", entry.totalBytes)
                    .put("startedAt", entry.startedAt)
                    .put("finishedAt", entry.finishedAt)
                    .put("success", entry.success)
                    .put("detail", entry.detail ?: "")
            )
        }
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
            .edit().putString(KEY_HISTORY, array.toString()).apply()
    }
}
